docs: update dev setup guide (#282)
fedora-infra/tahrir-api: README.rst

of interacting with the database. It allows for custom interactions with the
database without having to use the ``TahrirDatabase`` class.

Development
===========

System Requirements
-------------------

Before getting started, ensure your system meets the following requirements:

**Software Requirements**

- **Python**: 3.9 or higher 
- **Poetry**: Latest version for dependency management
- **Database**: PostgreSQL 9.6+ (recommended) or SQLite (basic development)
- **Git**: For version control

**Operating System Requirement**

- **Fedora Linux**: Latest non-EOL versions

Note: Other distributions may work, but are not officially supported. The maintainers may not be able to assist with platform-specific issues on non-Fedora systems.

Project Setup
-------------
1. Ensure that the necessary packages are installed:

> .. code-block:: bash

Reviewer: You forgot to add a new line.
<img width="1052" height="376" alt="Image" src="https://github.com/user-attachments/assets/cacb5da5-c5b3-4f54-9627-55883896e871" />
<img width="626" height="289" alt="Image" src="https://github.com/user-attachments/assets/a2d0bc9a-ca29-4b6f-9440-a4a015fea0ab" />

of interacting with the database. It allows for custom interactions with the
database without having to use the ``TahrirDatabase`` class.

Development
===========

System Requirements
-------------------

Before getting started, ensure your system meets the following requirements:

**Software Requirements**

> - **Python**: 3.9 or higher (tested with Python 3.9, 3.10, 3.11, and 3.12)

Reviewer: ```suggestion
- **Python**: 3.9 or higher
```
No need to mention about the testing versions in the requirement section.

> - **Poetry**: Latest version (for dependency management)

Reviewer: ```suggestion
- **Poetry**: Latest version for dependency management
```

> - **Database**: PostgreSQL 9.6+ or any SQLAlchemy-compatible database (MySQL, SQLite, etc.)

Reviewer: ```suggestion
- **Database**: PostgreSQL 9.6+ (recommended) or SQLite (basic development)
```

> - **Git**: For cloning and version control

Reviewer: ```suggestion
- **Git**: For version control
```

> **Operating Systems**

Reviewer: ```suggestion
**Operating System Requirement**
```

- Linux (Fedora/RedHat-based recommended)
- macOS

> - Windows (with WSL2 or similar)

Reviewer: ```suggestion
- **Fedora Linux**: Latest non-EOL versions

Note: Other distributions may work, but are not officially supported. The maintainers may not be able to assist with platform-specific issues on non-Fedora systems.
```

**Installation**

On RedHat-based systems (Fedora, CentOS, RHEL):

.. code-block:: bash

    $ sudo dnf install -y python3 python3-pip python3-devel gcc krb5-devel git postgresql

On Debian-based systems (Ubuntu, Debian):

.. code-block:: bash

    $ sudo apt-get install -y python3 python3-pip python3-dev gcc libkrb5-dev git postgresql

Project Setup

> -------------

Reviewer: ```suggestion
Project Setup
-------------

1. Ensure that the necessary packages are installed:
.. code-block:: bash

    $ sudo dnf install -y python3 python3-pip python3-devel poetry gcc krb5-devel git
```

The `postgresql` system package is not required as we are using the `postgresql` container.

> Clone the repository:

Reviewer: ```suggestion
2. Clone your fork to the local storage and make it your current working directory:
```

.. code-block:: bash

    $ git clone https://github.com/fedora-infra/tahrir-api.git
    $ cd tahrir-api

> Create and activate a virtual environment:

Reviewer: ```suggestion
3. Create a virtual environment and activate it for installing project dependencies:
```

.. code-block:: bash

    $ python3 -m venv venv
    $ source venv/bin/activate

> Install project dependencies:

Reviewer: ```suggestion
4. Install project dependencies:
```

.. code-block:: bash

>     $ pip install -e .

Reviewer: ```suggestion
    (venv )$ poetry check
    (venv )$ poetry install
```

Why are you using pip to install the project dependencies if you already have poetry for dependency management?

Database Setup
--------------

> Create a database directory and download the Tahrir snapshot:

Reviewer: ```suggestion
1. Create a directory for storing the Tahrir database snapshot:
```

.. code-block:: bash

    mkdir badges-database

>     cd badges-database

Reviewer: ```suggestion
    (venv )$ cd ..
    (venv )$ mkdir badges-database
    (venv )$ cd badges-database
```

> Download and extract the Tahrir database snapshot:

Reviewer: ```suggestion
2. Download and extract the Tahrir database snapshot:
```

.. code-block:: bash

    wget https://infrastructure.fedoraproject.org/infra/db-dumps/tahrir.dump.xz

>     unxz tahrir.dump.xz

Reviewer: ```suggestion
    (venv )$ wget https://infrastructure.fedoraproject.org/infra/db-dumps/tahrir.dump.xz
    (venv )$ unxz tahrir.dump.xz
```

> Create two new folders in the database folder for the data and the dump contents before moving the extracted dump over:

Reviewer: ```suggestion
3.  Create directories for persistent data and the database dump:
```

.. code-block:: bash

    mkdir data dump

>     mv tahrir.dump dump/

Reviewer: ```suggestion
    (venv )$ mkdir data dump
    (venv )$ mv tahrir.dump dump/
```

> Download the most recent postgres:15 from a provider like hub.docker.io before configuring and using the start command:

Reviewer: ```suggestion
4. Pull the PostgreSQL container image:
```

.. code-block:: bash

>     podman pull docker.io/library/postgres:15

Reviewer: ```suggestion
    (venv )$ podman pull docker.io/library/postgres:15
```

> Start the database container (be sure to rename the username and paths based on your development environment):

Reviewer: ```suggestion
5. Start the database container (be sure to rename the username and paths based on your development environment):
```

.. code-block:: bash

>     podman run \

Reviewer: ```suggestion
    (venv )$ podman run \
```

        --name badges-database \
        --env POSTGRES_USER=badgesdb \
        --env POSTGRES_PASSWORD=badgesdb \
        --env POSTGRES_DB=badgesdb \
        --env PGDATA=/var/lib/postgresql/data/pgdata \
        --volume /home/username/Projects/badges-database/data:/var/lib/postgresql/data:Z \
        --volume /home/username/Projects/badges-database/dump:/badgesdb:Z \
        --publish 5432:5432 \
        --restart unless-stopped \
        --detach docker.io/library/postgres:15

> Once the database container has started, log in to the interactive shell using the password to begin importing the extracted dump:

Reviewer: ```suggestion
6. Once the database container has started, log in to the interactive shell using the password to begin importing the extracted dump:
```

.. code-block:: bash

>     podman exec -ti badges-database psql --username badgesdb --password

Reviewer: ```suggestion
    (venv )$ podman exec -ti badges-database psql --username badgesdb --password
```

> Then execute the following SQL commands:

Reviewer: ```suggestion
7. Execute the following SQL commands to create roles, load the database dump, and grant the needed access:
```

.. code-block:: sql

    badgesdb=# create role "tahrir" with inherit nocreatedb nocreaterole noreplication nosuperuser valid until 'infinity' login password 'tahrir';
    badgesdb=# create role "tahrir-readonly" with nocreatedb inherit nocreaterole noreplication nosuperuser valid until 'infinity' login password 'tahrir-readonly';
    badgesdb=# \i badgesdb/tahrir.dump
    badgesdb=# grant connect on database tahrir to "tahrir";
    badgesdb=# grant all on all tables in schema public to "tahrir";
    badgesdb=# grant all on all sequences in schema public to "tahrir";
    badgesdb=# revoke create on schema public from public;

> Exit out of the interactive shell and log back in using the newly created credentials:

Reviewer: ```suggestion
8. Exit out of the interactive shell and log back in using the newly created credentials:
```

.. code-block:: bash

>     podman exec -ti badges-database psql --username tahrir --password

Reviewer: ```suggestion
    (venv )$ podman exec -ti badges-database psql --username tahrir --password
```

> Verify the database setup:

Reviewer: ```suggestion
9. Verify the database setup:
```

.. code-block:: sql

>     tahrir=> \dt+

Reviewer: ```suggestion
    (venv )$ tahrir=> \dt+
```

Happy hacking!    

Run the tests
-------------

You can run the tests with ``tox``:

.. code-block:: bash

    $ pip install tox poetry

>     $ tox

Reviewer: ```suggestion
Run the tests
-------------

1. Download and install ``tox``:

.. code-block:: bash

    (venv) $ cd ../tahrir-api
    (venv) $ sudo dnf install tox

2. Run the tests with ``tox``:

.. code-block:: bash

    (venv) $ tox
```

Creating a Badge
================

This is an example of creating a badge via Tahrir-API:

.. code-block:: python

    from tahrir_api.dbapi import TahrirDatabase

    db = TahrirDatabase('backend://badges:badgesareawesome@localhost/badges')

    origin = 'http://foss.rit.edu/badges'
    issuer_name = 'FOSS@RIT'
    org = 'http://foss.rit.edu'
    contact = 'foss@rit.edu'

    issuer_id = db.add_issuer(origin, issuer_name, org, contact)

    badge_name = 'fossbox'
    image = 'http://foss.rit.edu/files/fossboxbadge.png'
    desc = 'Welcome to the FOSSBox. A member is you!'
    criteria = 'http://foss.rit.edu'

    db.add_badge(badge_name, image, desc, criteria, issuer_id)

Awarding a Badge
================

This is an example of awarding a badge via Tahrir-API:

.. code-block:: python

    from tahrir_api.dbapi import TahrirDatabase

    db = TahrirDatabase('backend://badges:badgesareawesome@localhost/badges')

    badge_id = 'fossbox'
    person_email = 'person@email.com'
    issued_on = None

    db.add_person(person_email)

>     db.add_assertion(badge_id, person_email, issued_on)

Reviewer: I believe these examples can be removed and added to some other document. @gridhead & @abompard, I'd like to know your opinions.
